### ISSUE TITLE
ps_specific_price_priority unique key

### ISSUE THREAD

**Issue by musicpanda** [PrestaShop/PrestaShop#37725]
### Prerequisites

- [x] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md).
- [x] I have already [searched in existing issues](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3ABug) and found no previous report of this bug.

### Describe the bug and add attachments

Trying to fit an older database into the PS 8.2 database structure I found something strange with the ps_specific_price_priority table. This table has a unique index for id_product. 

This is rather strange. The table has also a unique key for the combination of id_product and id_specific_price_priority. It also seems to beat the reason of existence of the table of ranking priorities for a product.

I am not sure when this was introduced but I don't see a clause in autoupgrade to fit existing shops into this straitjacket.

### Expected behavior

The id_product index should not be unique

### Steps to reproduce

Look at the structure of the ps_specific_price_priority table under PS 8.2

### PrestaShop version(s) where the bug happened

8.2

**Comment by paulnoelcholot** [comment]
Hello @musicpanda,

I try to reproduice your issues but as you can see, I can't : 

![Image](https://github.com/user-attachments/assets/700a330e-7018-493d-b4cf-f1e10b108604)

Can you tell me how can I haven't ps_specific_price_priority empty?

Thanks for you feedback!

**Comment by musicpanda** [comment]
Hello Paul,

Your answer puzzles me. 

The attached picture is what I mean:
![Image](https://github.com/user-attachments/assets/98879c17-b1e3-47da-aa22-9287019d581a)

**Comment by paulnoelcholot** [comment]
Hello @musicpanda ,

Thanks for your feedback, now I reproduce the issue with PrestaShop version 8.2.x, I'll add this to the backlog so it can be fixed.

Please be aware that some issues might take a very long time to be resolved. If this one is important to you and you cannot wait for it to be fixed on the project’s own time, we strongly suggest you consider [contacting a professional](https://www.prestashop-project.org/support/) to help you. 

If you fix the issue on your end, please [contribute it back to the project](https://devdocs.prestashop.com/1.7/contribute/contribute-pull-requests/). Remember that the more people contribute, the better PrestaShop becomes for everyone.

Thank you

![Image](https://github.com/user-attachments/assets/012100aa-c44c-46e1-9c09-a1b037f244f4)

**Comment by Hlavtox** [comment]
@musicpanda Is this really a bug? I see more of a duplicate key more than anything. Does it affect the funcionality. If I am not wrong, there can be only one specific price priority on one product.

So yes, one key would be sufficient, but I think it's not visible in the software.

Ping @jolelievre @kpodemski

**Comment by musicpanda** [comment]
I always thought that this table was to determine which discount is applied first if an ordered product has more than one discount. For example a group discount, a category discount and a product discount.

**Comment by Hlavtox** [comment]
@musicpanda Yes, this is how it should work, but only once per product, not between specific prices.

The priority between specific prices is just broken and people dont seem to care..

**Comment by musicpanda** [comment]
> [@musicpanda](https://github.com/musicpanda) Yes, this is how it should work, but only once per product, not between specific prices.
> 
> The priority between specific prices is just broken and people dont seem to care..

That reminds me of the Advanced Stock Management. There we got the same story: it doesn't work well and few people use it so let us delete it. It is not a very ambitious approach and it makes me wonder how much will be left five years from now.

**Comment by Hlavtox** [comment]
@musicpanda I have checked it again, UNIQUE id_product index has a purpose, to prevent there being two records for single product ID.

If I delete it, I can create following data, which is wrong.

![Image](https://github.com/user-attachments/assets/160aa046-7f65-4544-a181-dd161ac66591)

**Comment by musicpanda** [comment]
> [@musicpanda](https://github.com/musicpanda) I have checked it again, UNIQUE id_product index has a purpose, to prevent there being two records for single product ID.
> 
> If I delete it, I can create following data, which is wrong.
> 
> ![Image](https://github.com/user-attachments/assets/160aa046-7f65-4544-a181-dd161ac66591)

Your reaction puzzles me. At one time you admit that that the software is broken and that the unique id_product index is a band aid to avoid misleading users. And then the next moment you claim that it was always meant to this way.

I see quite a lot of shops setting this feature. I assume that the lack of complaints is the product of the circumstances. Small shops will seldom have multiple discounts and when they have they usually won't bother to check the math that the setting works. And larger shops will prefer buying a module rather than fielding a bug report that is unlikely to be soon fixed.

**Comment by boo-code** [comment]
@Hlavtox's conclusion is right, and there is a mechanism behind it rather than just the shape of the data:
the unique key on `id_product` is what the write depends on. `SpecificPrice::setSpecificPriority()` saves
with an upsert:

```sql
INSERT INTO `PREFIX_specific_price_priority` (`id_product`, `priority`)
VALUES (..., '...')
ON DUPLICATE KEY UPDATE `priority` = '...'
```

`ON DUPLICATE KEY UPDATE` has nothing to match on without that index, so every save becomes an insert.
Measured on 9.2 by copying the table and repeating the same two upserts:

```
with the unique index      1 row for the product
without the unique index   3 rows for the product
```

It does not visibly break the priority, because the read is
`ORDER BY id_specific_price_priority DESC` with `getValue()` and so still picks the newest row - but the
table grows by one row per save and the one-priority-per-product invariant is gone. So dropping the index
is the opposite of a fix, and the issue is worth closing before someone acts on the `Waiting for dev`
label.

On the second index in the report - there is no unique key on `(id_product, id_specific_price_priority)`.
That is the PRIMARY KEY, declared the other way round:

```sql
PRIMARY KEY (`id_specific_price_priority`, `id_product`),
UNIQUE KEY `id_product` (`id_product`)
```

`id_specific_price_priority` is `AUTO_INCREMENT` and therefore already unique on its own, so adding
`id_product` to the primary key adds no uniqueness - and in InnoDB the primary key is carried inside every
secondary index, so the `id_product` index stores both columns. That is a real redundancy, and it is the
grain of truth in the report. It is also worth almost nothing in practice: the table holds at most one row
per product that has a custom priority, so this is a handful of bytes on a table that is usually near
empty. Not worth a schema change on a patch version by itself - but if this table is ever touched for
another reason, `PRIMARY KEY (id_specific_price_priority)` is the correct declaration.
